src/controller/CHIPDeviceController.cpp

```diff
@@ -528,6 +533,52 @@ void DeviceController::OnValue(const char * key, const char * value) {}
 
 void DeviceController::OnStatus(const char * key, Operation op, CHIP_ERROR err) {}
 
+#if CHIP_ENABLE_MDNS
+
+void DeviceController::OnNodeIdResolved(CHIP_ERROR error, uint64_t nodeId, const Mdns::ResolvedNodeData & nodeData)
+{
+    Device * device   = nullptr;
+    bool addressValid = false;
+
+    SuccessOrExit(error);
+    SuccessOrExit(error = GetDevice(nodeId, &device));
+
+#if INET_CONFIG_ENABLE_IPV4
+    addressValid |= (nodeData.mAddressType == Inet::kIPAddressType_IPv4);
+#endif
+    addressValid |= (nodeData.mAddressType == Inet::kIPAddressType_IPv6);
+    VerifyOrExit(addressValid, error = CHIP_ERROR_INVALID_ADDRESS);
+
+    device->UpdateAddress(nodeData.mAddress, nodeData.mPort);
+
+exit:
+    if (mDiscoveryDelegate != nullptr)
+    {
+        mDiscoveryDelegate->OnDiscoveryComplete(nodeId, error);
+    }
+}
+
+CHIP_ERROR DeviceController::DiscoverDevice(uint64_t fabricId, NodeId nodeId)
+{
+    return Mdns::Resolver::Instance().ResolveNodeId(nodeId, fabricId, Inet::kIPAddressType_Any);
+}
+
+#else // CHIP_ENABLE_MDNS
+
+void DeviceController::OnNodeIdResolved(CHIP_ERROR error, uint64_t nodeId, const Mdns::ResolvedNodeData & nodeData)
+{
+    IgnoreUnusedVariable(error);
+    IgnoreUnusedVariable(nodeId);
+    IgnoreUnusedVariable(nodeData);
+}
+
+CHIP_ERROR DeviceController::DiscoverDevice(uint64_t fabricId, NodeId nodeId)
```

Comment:
I am concerned that we make the device controller do too much - this couples the device controller with global state regarding resolution.

How about having discovery be independent of the controller?

For the long term, I imagine we want to cache these resolutions so that we do not require to rediscover nodes on every operation. For that purpose, a separate object whose purpose is to do cached discovery seems desirable. I would also make it easier to follow the 'only one delegate' support in the Resolver.

Reply:
Yeah, I wasn't sure what would be the final usage of this method and where to put it. According to the spec we should perform the discovery when the address is unknown or a message dispatch failed (I guess that will be possible to verify once CRMP is used), so I'm not sure if the global cache will be necessary. I put the method here as I believed the controller will know the fabric ID and the method will be needed after commissioning, but I can move it somewhere else. Do you think some singleton class like `chip::Controller::DiscoveryManager` would be a better place?

Reply:
I now perfer to have a simple `ResolveNode` and leave the caching mechanism to the implementation. Avahi and mDNSResponder are shipped with caching and TTL managmement. Caching will be hard on these systems since they do not expose the TTL to users. On minimal mDNS we may maintain our own cache.

The api we expose to other modules can be a simple `ResolveNodeId` but I'd also like it to be a separate class other than the `DeviceController`.

Reply:
> I perfer to have a simple `ResolveNode` and leave the caching mechanism to the implementation. Avahi and mDNSResponder are shipped with caching and TTL managmement. Another cache in CHIP will be hard on these platforms since they do not expose the TTL to us. On minimal mDNS we may maintain our own cache.

I have to agree.  Building caching into the CHIP stack seems like it pulls way more into the sdk than we need to.  If we have an mDNS resolver that doesn't cache, it seems like effort would be better expended to fix that.

Reply:
I created a separate `DeviceAddressUpdater`, but it can't be completely independent from the controller since it must update addresses in device objects owned by the controller. However, it should be a bit more separated now.